src/Compilers/Core/Rebuild/CompilationOptionsReader.cs

```diff
@@ -271,35 +284,75 @@ public ImmutableArray<MetadataReferenceInfo> GetMetadataReferences()
             return result;
         }
 
-        public ImmutableArray<SourceFileInfo> GetSourceFileInfos(Encoding encoding)
+        public (ImmutableArray<SyntaxTree> SyntaxTrees, ImmutableArray<MetadataReference> MetadataReferences) ResolveArtifacts(
+            IRebuildArtifactResolver resolver,
+            Func<string, SourceText, SyntaxTree> createSyntaxTreeFunc)
         {
-            var sourceFileCount = int.Parse(
-                GetMetadataCompilationOptions()
-                    .GetUniqueOption(CompilationOptionNames.SourceFileCount));
+            var syntaxTrees = ResolveSyntaxTrees();
+            var metadataReferences = ResolveMetadataReferences();
+            return (syntaxTrees, metadataReferences);
 
-            var builder = ImmutableArray.CreateBuilder<SourceFileInfo>(sourceFileCount);
-            foreach (var documentHandle in PdbReader.Documents.Take(sourceFileCount))
+            ImmutableArray<SyntaxTree> ResolveSyntaxTrees()
             {
-                var document = PdbReader.GetDocument(documentHandle);
-                var name = PdbReader.GetString(document.Name);
-
-                var hashAlgorithmGuid = PdbReader.GetGuid(document.HashAlgorithm);
-                var hashAlgorithm =
-                    hashAlgorithmGuid == HashAlgorithmSha1 ? SourceHashAlgorithm.Sha1
-                    : hashAlgorithmGuid == HashAlgorithmSha256 ? SourceHashAlgorithm.Sha256
-                    : SourceHashAlgorithm.None;
+                var sourceFileCount = GetSourceFileCount();
+                var builder = ImmutableArray.CreateBuilder<SyntaxTree>(sourceFileCount);
+                foreach (var tuple in GetSourceTextInfoCore())
```

Comment:
>var tuple [](start = 25, length = 9)

Consider using `var (document, sourceText) in GetSourceTextInfoCore()`.

Reply:
IIRC, something about the target frameworks prevents us from actually deconstructing tuples.

Elsewhere in the Rebuild layer we have this:

https://github.com/dotnet/roslyn/blob/e5fa846d30c4ed4bf5804771342cb86f894d4330/src/Tools/BuildValidator/Program.cs#L400

Perhaps a follow-up issue to dig into why we can't deconstruct these. I don't recall having this problem in the compiler, but perhaps it is there too.

Reply:
The only issue we should have is deconstructing `KeyValuePair<TKey, TValue>` because the `Deconstruct` method wasn't added in `net472`. All other deconstruct should work in our code base. Think we have a place where we added an extension method for KVP so it would work too.